### PR TITLE
Reworked the math evaluation to work on u64, i64 or f64

### PR DESCRIPTION
**Pull request by vdagonneau**
#301 This is a basic fix for the problem. It just solves it. It also now panics on NaN operations (divide by 0), I am not sure if it is the expected behavior. I also haven't provided tests yet.

### CODE REVIEW

**Comment by Keats** [comment]
Hmm I don't think panicking on NaN makes sense, we don't want any template rendering to panic. I think the current behaviour of returning NaN makes sense as it's not an error of the template engine directly - you could get a NaN purely from user data and it would make sense to display NaN rather than not being able to render the template

File: src/renderer/mod.rs [hunk 1]
```diff
@@ -253,34 +253,106 @@ impl<'a> Renderer<'a> {
...
+                            let rr = r.as_i64().unwrap();
+                            Ok(Number::from(ll * rr))
+                        } else if l.is_u64() && r.is_u64() {
+                            let ll = l.as_u64().unwrap();
+                            let rr = r.as_u64().unwrap();
+                            Ok(Number::from(ll * rr))
+                        } else {
+                            let ll = l.as_f64().unwrap();
+                            let rr = r.as_f64().unwrap();
+			    Ok(Number::from_f64(ll * rr).unwrap())
```
Comment (Keats) [inline src/renderer/mod.rs]: wrong indent

File: src/renderer/mod.rs [hunk 2]
```diff
@@ -253,34 +253,106 @@ impl<'a> Renderer<'a> {
...
+                    MathOperator::Div => {
+                        let ll = l.as_f64().unwrap();
+                        let rr = r.as_f64().unwrap();
+			let res = ll / rr;
+			if res.is_nan() {
+			    Err(From::from("NaN"))
+			} else {
+                            Ok(Number::from_f64(res).unwrap())
+			}
+                    },
```
Comment (Keats) [inline src/renderer/mod.rs]: wrong indent as well

File: src/renderer/mod.rs [hunk 3]
```diff
@@ -253,34 +253,106 @@ impl<'a> Renderer<'a> {
...
+                    MathOperator::Add => { 
+                        if l.is_i64() && r.is_i64() {
+                            let ll = l.as_i64().unwrap();
+                            let rr = r.as_i64().unwrap();
+                            Ok(Number::from(ll + rr))
+                        } else if l.is_u64() && r.is_u64() {
+                            let ll = l.as_u64().unwrap();
+                            let rr = r.as_u64().unwrap();
+                            Ok(Number::from(ll + rr))
```
Comment (Keats) [inline src/renderer/mod.rs]: I believe in all cases other than `float` we can unwrap numbers as `i64` no?

Comment (vdagonneau) [reply src/renderer/mod.rs]: Well, I'm pretty sure you can't cast a u64 to i64.

For example, if the input number is std::i64::MAX + 1, it is a valid u64 but definitely not the same value as an i64.

The JSON Number type from serde stores either the f64, u64 or i64 based on the value provided. We need to get the correct one. I agree it is very verbose and I hope we can improve that, I'm open to suggestion.

For reference, the implementation of Number: https://github.com/serde-rs/json/blob/master/src/number.rs

**Comment by vdagonneau** [comment]
I'll correct the indentation and I am also on my way to correct the two tests that are failing.

**Comment by Keats** [comment]
Last thing I forgot to ask, can you add a test for the `range` issue you had?

**Comment by vdagonneau** [comment]
It should be good to go now.

**Comment by Keats** [comment]
Thanks!

**Pull request merged**